ipa-core/src/protocol/ipa_prf/shuffle/malicious.rs

```diff
@@ -531,4 +782,116 @@ mod tests {
+    fn fail_under_bit_flip_attack_on_x2() {
+        const RECORD_AMOUNT: usize = 10;
+
+        run(move || async move {
+            let mut rng = thread_rng();
+            let mut config = TestWorldConfig::default();
+            config.stream_interceptor =
+                MaliciousHelper::new(Role::H1, config.role_assignment(), interceptor_h1_to_h2);
+
+            let world = TestWorld::new_with(config);
+            let records = (0..RECORD_AMOUNT).map(|_| rng.gen()).collect::<Vec<BA32>>();
+            let [_, h2, _] = world
+                .semi_honest(records.into_iter(), |ctx, shares| async move {
+                    malicious_shuffle::<_, BA32, BA64, _>(ctx, shares).await
+                })
+                .await;
+
+            let _ = h2.unwrap();
+        });
+    }
+
+    /// This test checks that the malicious sort fails
+    /// under a simple bit flip attack by H2.
+    ///
+    /// `y1` will be inconsistent which is checked by `H1`.
+    #[test]
+    #[should_panic(expected = "Y1 is inconsistent")]
+    fn fail_under_bit_flip_attack_on_y1() {
+        const RECORD_AMOUNT: usize = 10;
+
+        run(move || async move {
+            let mut rng = thread_rng();
+            let mut config = TestWorldConfig::default();
+            config.stream_interceptor =
+                MaliciousHelper::new(Role::H2, config.role_assignment(), interceptor_h2_to_h3);
+
+            let world = TestWorld::new_with(config);
+            let records = (0..RECORD_AMOUNT).map(|_| rng.gen()).collect::<Vec<BA32>>();
+            let [h1, _, _] = world
+                .malicious(records.into_iter(), |ctx, shares| async move {
+                    malicious_shuffle::<_, BA32, BA64, _>(ctx, shares).await
+                })
+                .await;
+            let _ = h1.unwrap();
+        });
+    }
+
+    /// This test checks that the malicious sort fails
+    /// under a simple bit flip attack by H3.
+    ///
+    /// `c` from `H2` will be inconsistent
+    /// which is checked by `H1`.
+    #[test]
+    #[should_panic(expected = "C from H2 is inconsistent")]
+    fn fail_under_bit_flip_attack_on_c() {
+        const RECORD_AMOUNT: usize = 10;
+
+        run(move || async move {
+            let mut rng = thread_rng();
+            let mut config = TestWorldConfig::default();
+            config.stream_interceptor =
+                MaliciousHelper::new(Role::H3, config.role_assignment(), interceptor_h3_to_h2);
+
+            let world = TestWorld::new_with(config);
+            let records = (0..RECORD_AMOUNT).map(|_| rng.gen()).collect::<Vec<BA32>>();
+            let [h1, h2, _] = world
+                .semi_honest(records.into_iter(), |ctx, shares| async move {
+                    malicious_shuffle::<_, BA32, BA64, _>(ctx, shares).await
+                })
+                .await;
+
+            // x2 should be consistent with y2
+            let _ = h2.unwrap();
+
+            // but this should fail
+            let _ = h1.unwrap();
```

Comment:
I believe the `let _` are unnecessary here?

The way the test is written is fine, but FYI, you could also use `assert!(matches!(...))` to verify the error without `should_panic`.